### DISCUSSION TITLE
Critical bug in NextJS v14.2.7 - renders page source code instead of components

### DISCUSSION THREAD

**Discussion by billypap1** [vercel/next.js#69491] [Help]
### Summary

I have created my app in NextJS (v14.2.7) with App Router and I have deployed it (dockerized) in App Platform Digital Ocean.

At random times it shows in the browser the following instead of the page:
<img width="813" alt="Screenshot at Aug 30 13-09-23" src="https://github.com/user-attachments/assets/c844ed21-9251-415f-a5f9-66489ee95a7d">

It's like it prints the page source code instead of running it. It's the same bug as it was in v13.4.5
I don't receive any error in Sentry and there isn't any issue in dev mode.

**Comment by icyJoseph** [comment]
IIRC this is an issue with how your server is configured. It is caching requests with an RSC query (lowercased), as if they were the actual path.

You need to make your CDN, or however your stack is built, to differentiate requests with their query parameters.

This has been reported and fixed before. I can try to post a few links to issues on this repository.

**Reply by billypap1** [reply]
@icyJoseph  I use Cloudflare with NextJS, what page rule should I add in order to differentiate requests with their query parameters?

**Reply by icyJoseph** [reply]
I searched here, https://developers.cloudflare.com/cache/how-to/set-caching-levels/ and the default behavior seems to do as I am describing, could you verify your configuration?

Also, when you do get, the RSC payload, do you see headers indicating cache HIT?

**Reply by icyJoseph** [reply]
I must also add, that what you are seeing is not page source code. It is serialized VDOM, ska, RSC string, that describes your UI. That is, the result of rendering your UI tree.

**Reply by billypap1** [reply]
I get the following headers for _rsc: 

```
age:1157
alt-svc: h3=":443"; ma=86400
cache-control:s-maxage=31536000, stale-while-revalidate
cf-cache-status: DYNAMIC
cf-ray: 8bb48764ce3eca21-OTP
x-frame-options: DENY
x-next-i18n-router-locale: en
x-nextjs-cache: HIT
server: cloudflare
vary: RSC, Next-Router-State-Tree, Next-Router-Prefetch, Accept-Encoding
```
The settings in cloudflare for cache is in Standard mode.

**Reply by icyJoseph** [reply]
And for a normal request, without RSC, that still shows the RSC payload (like in your screenshot), what headers do you get?

**Reply by billypap1** [reply]
This is from page chunk.js: 
```
age:825
alt-svc: h3=":443"; ma=86400
cache-control: public, max-age=31536000, immutable
cf-cache-status: HIT
cf-ray: 8ba3eaa19ae7051b-OTP
server: cloudflare
vary: Accept-Encoding
x-content-type-options: nosniff
x-do-orig-status: 200
x-frame-options: DENY
```

and this from route page: 

```
alt-svc: h3=":443"; ma=86400
cache-control: s-maxage=31536000, stale-while-revalidate
cf-cache-status: DYNAMIC
cf-ray: 8bb4990a4dbfc9f3-OTP
content-encoding: br
x-do-orig-status:200
x-frame-options:DENY
x-middleware-rewrite: /en
x-next-i18n-router-locale: en
x-nextjs-cache:HIT
x-powered-by:Next.js
```

According to the docs for Digital Ocean App Platform, the app platform use Cloudflare CDN ( that I don't have access) by default. I use also my own cloudflare CDN on top of App Platform's CDN.
Is this maybe the issue? The cache is not synced?

**Reply by icyJoseph** [reply]
Yeah, I think something is up with cloudflare.

Could you see if this post/thread helps you? https://github.com/vercel/next.js/issues/48569#issuecomment-1518887170

**Reply by billypap1** [reply]
Unfortunately I don't have access to Digital Ocean App Platform's Cloudflare CDN. I turned off my cloudflare CDN and it's active only Digital Ocean's. 

The new headers I get are the same but with this change: 
cf-cache-status: HIT and it's not DYNAMIC.

**Reply by billypap1** [reply]
@icyJoseph I tested only with Digital Ocean's Cloudflare CDN and I get the same issue. Maybe app platform is not suitable for NextJS?

**Reply by icyJoseph** [reply]
Read here maybe? https://github.com/vercel/next.js/issues/49130

I'd be surprised to learn that, this is not a known fixable issue

**Reply by billypap1** [reply]
So is this entirely a CDN issue and not internal to NextJS? The only solution is to migrate to an other provider.

**Reply by Hyferion** [reply]
I'm having the same issue with DO App Platform. I contacted support and they told me they do not have a solution for it.

Is there anything new on this, or is the only solution to switch provider?
